carla/recourse_methods/catalog/face/library/face_method.py

```diff
@@ -0,0 +1,229 @@
+# import helpers
+import numpy as np
+
+# import Dijkstra's shortest path algorithm
+from scipy.sparse import csgraph, csr_matrix
+
+# import graph building methods
+from sklearn.neighbors import kneighbors_graph, radius_neighbors_graph
+
+
+def graph_search(
+    data,
+    index,
+    keys_mutable,
+    keys_immutable,
+    continuous_cols,
+    binary_cols,
+    model,
+    n_neighbors=50,
+    p_norm=2,
+    mode="knn",
+    frac=0.4,
+    radius=0.25,
+):
+    # This is our own implementation of these methods since the authors did not provide implementations online,
+    # neither did they respond to our emails
+    # We implemented the eps radius graph and the knn graph methods. The density graph will be added in the future.
+
+    # This one implements the FACE method from
+    # Rafael Poyiadzi et al (2020), "FACE: Feasible and Actionable Counterfactual Explanations",
+    # Conference on AI, Ethics & Accountability (AIES, 2020)
+
+    """
+    :param data: df
+    :param n_neighbors: int > 0; number of neighbors when constructing knn graph
+    :param step: float > 0; step_size for growing spheres
+    :param mode: str; either 'knn' or 'epsilon'
+    :param model: classification model (either tf keras, pytorch or sklearn)
+    :param p_norm: float=>1; denotes the norm (classical: 1 or 2)
+    :param frac: float 0 < number =< 1; fraction of data for which we compute the graph; if frac = 1, and data set large, then compute long
+    :param keys_mutable: list; list of input names we can search over
+    :param keys_immutable: list; list of input names that may not be searched over
+    :param radius: float > 0; parameter for epsilon density graph
+    :return: candidate_counterfactual_star: np array (min. cost counterfactual explanation)
+    """
+
+    # Divide data in 'mutable' and 'non-mutable'
+    # In particular, divide data in 'mutable & binary' and 'mutable and continuous'
+
+    # Choose a subset of data for computational efficiency
+    number_samples = np.int(np.rint(frac * data.values.shape[0]))
+    chosen_indeces = np.random.choice(
+        np.arange(0, index).tolist()
+        + np.arange(index + 1, data.values.shape[0]).tolist(),
+        replace=False,
+        size=number_samples,
+    )
+    chosen_indeces = [
+        index
+    ] + chosen_indeces.tolist()  # make sure sample under consideration included
+    data = data.iloc[chosen_indeces]
+    data = data.sort_index()
+
+    # ADD CONSTRAINTS by immutable inputs into adjacency matrix
+    # if element in adjacency matrix 0, then it cannot be reached
+    # this ensures that paths only take same sex / same race / ... etc. routes
+    # TODO: CANNOT deal with continuous constraints (e.g. age) yet
+    # TODO: However, authors also do not mention how to go about it
```

Comment:
Also unsure if we should keep all this comments here. Should we create issues for the TODOs instead if relevant or put further documenation of the algorithm in a e.g. docs/face.md if needed?

Reply:
Lets create an "improve face method" issue and add these todos there. We don't have to put it on the project board to tackle immediatly

Reply:
👍